lib/posix/pthread.c

```diff
@@ -198,38 +198,59 @@ void __z_pthread_cleanup_pop(int execute)
 	}
 }
 
-static bool is_posix_policy_prio_valid(uint32_t priority, int policy)
+static bool is_posix_policy_prio_valid(int priority, int policy)
 {
-	if (priority >= sched_get_priority_min(policy) &&
-	    priority <= sched_get_priority_max(policy)) {
-		return true;
+	int priority_limit = sched_get_priority_min(policy);
+
+	if (priority_limit >= 0 && priority >= priority_limit) {
+		priority_limit = sched_get_priority_max(policy);
+		if (priority_limit >= 0 && priority <= priority_limit) {
+			return true;
+		}
 	}
 
 	LOG_ERR("Invalid piority %d and / or policy %d", priority, policy);
 
 	return false;
 }
 
-static uint32_t zephyr_to_posix_priority(int32_t z_prio, int *policy)
+static int zephyr_to_posix_priority(int32_t z_prio, int *policy)
 {
 	if (z_prio < 0) {
-		__ASSERT_NO_MSG(z_prio < CONFIG_NUM_COOP_PRIORITIES);
+		/* COOP: highest [-CONFIG_NUM_COOP_PRIORITIES, -1] low */
+		__ASSERT_NO_MSG((CONFIG_NUM_COOP_PRIORITIES > 0) &&
+				(z_prio < -CONFIG_NUM_COOP_PRIORITIES));
+	} else {
+		/* PREEMPT: high [0, CONFIG_NUM_PREEMPT_PRIORITIES -1] lowest */
+		__ASSERT_NO_MSG((CONFIG_NUM_PREEMPT_PRIORITIES > 0) &&
+				(z_prio < CONFIG_NUM_PREEMPT_PRIORITIES));
+	}
+
+	if (IS_ENABLED(CONFIG_PREEMPT_ENABLED) &&
+	    IS_ENABLED(CONFIG_COOP_ENABLED)) {
+		*policy = SCHED_OTHER;
 	} else {
-		__ASSERT_NO_MSG(z_prio < CONFIG_NUM_PREEMPT_PRIORITIES);
+		*policy = (z_prio < 0) ? SCHED_FIFO : SCHED_RR;
 	}
 
-	*policy = (z_prio < 0) ? SCHED_FIFO : SCHED_RR;
-	return ZEPHYR_TO_POSIX_PRIORITY(z_prio);
+	return ZEPHYR_TO_POSIX_PRIORITY(z_prio) +
+		(IS_ENABLED(CONFIG_PREEMPT_ENABLED) ?
+			CONFIG_NUM_PREEMPT_PRIORITIES : 0);
```

Comment:
With the additional information that POSIX priorities must be >= 0, this probably needs a bit of a rework.

Ping me on Discord if you need any 1:1

Reply:
Hmm... I think this will be a lot easier to do once #67223 is done.

The problem:

without having a copy of the thread policy built-in to `struct posix_thread`, it's impossible to know, with confidence, whether someone using the `k_thread` API has modified the policy of a pthread. We can only hope, currently, that is not the case. So there is a chance that pthread and k_thread priority changes can introduce undetectable changes in scheduling policy, which is... not great.

At least with #67223 it will be possible to detect if a thread priority has been modified behind our backs.